psycopg3/transaction.py

```diff
@@ -50,29 +88,48 @@ def __exit__(
         exc_tb: Optional[TracebackType],
     ) -> None:
         with self._conn.lock:
-            if exc_type is None:
+            if exc_type is None and not self.force_rollback:
+                # Commit changes made in the transaction context
                 if self._savepoint_name:
                     self._conn._savepoints.pop()
                     # TODO: Add test for this assert
                     # assert self._conn._savepoints.pop() == self._savepoint
-                    self._conn._exec_command(
-                        b"release savepoint " + self._savepoint_name
+                    self._log_and_exec_command(
+                        b"RELEASE SAVEPOINT " + self._savepoint_name
                     )
                 if self._outer_transaction:
                     # TODO: Add test for this assert
                     # assert len(self._conn._savepoints) == 0
-                    self._conn._exec_command(b"commit")
+                    self._log_and_exec_command(b"COMMIT")
                     self._conn._savepoints = None
             else:
+                # Rollback changes made in the transaction context
+                if exc_type is Rollback:
+                    _log.debug(
+                        f"{self._conn}: Explicit rollback from: ",
+                        exc_info=True,
+                    )
+
                 if self._savepoint_name:
                     self._conn._savepoints.pop()
                     # TODO: Add test for this assert
                     # assert self._conn._savepoints.pop() == self._savepoint
-                    self._conn._exec_command(
-                        b"rollback to savepoint " + self._savepoint_name
+                    self._log_and_exec_command(
+                        b"ROLLBACK TO SAVEPOINT " + self._savepoint_name
                     )
                 if self._outer_transaction:
```

Comment:
Isn't this an `elif`?

Reply:
Both blocks should execute in what I'm calling "use case 2", when there is no outer transaction in progress yet, _and_ the user has specified a particular `savepoint_name` -- here the context will "2. Begin an outer transaction and create a savepoint (if one is named)"

We could optimise slightly because a `ROLLBACK TO SAVEPOINT` (or `RELEASE SAVEPOINT`) immediately followed by a `ROLLBACK` (or `COMMIT`) is redundant, but it will complicate the code (we'd still need to do the correct bookkeeping with the savepoint stack), and we've already said we don't care about the performance hit of issuing both `BEGIN` and `SAVEPOINT` commands (only for use case 2).

Reply:
Also, your question highlighted that there was no test coverage for this, which I've now added. Thanks :-)